File: newconting/FluxoCA/estado.c
/*----------------------estado.c------------------------------------------------
FAZ CHAMADAS RELATIVAS AS ROTINAS DE ZOLLENKOPF
FONTE DESTE ALGORITMO: Fluxograma da pag. 95 do Livro do Alcir Monticelli:
Fluxo de Carga em Redes de Energia Elétrica, Editora Edgard Blucher Ltda., 1983.
-------------------------------------------------------------*/
/*------------------------------------------------------
  Bibliotecas e arquivos do tipo include do compilador
------------------------------------------------------*/
#ifdef WIN_32
#include <windows.h>
#endif
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <math.h>

/*---------------------------------------
  Arquivo do tipo include do aplicativo
----------------------------------------*/
#include "global.h"   

/*---------------------Prototipos de funcoes----------------------------------*/

void matrizes(double[],double[]);
void elim(double valores[],int nn);
void redu(double[]);
void potencia(void);
void solu(double xce[],double b[]);
void controle(double[],double[]);

/*-------------------------------------------------------------------------
  Declaracoes de lista passada via parametro (fatores L,D,U de B'')
  -----------------------------------------------------------------------*/
double cleb[maxnzero];

/*----------------------Inicio de estado() -----------------------------------*/

void estado(double ce[],double cle[]){
	double dx1[nbmax],dx2[nbmax],dpmax,dqmax,dpa,dqa,qg,aux;
	int k,kp,kq,inic,i,flagq;
	double vmi,vma,ten;
	int kmi,kma;
	
	/*-----------Inicializa contadores e chaves:------------*/
	nitea = 0;        // contador de iteracoes ativas
	niter = 0;        // contador de iteracoes reativas
	ntot  = 0;
	lim   = 1600;
	kp    = 1;        // chaves
	kq    = 1;
	
	/*--------- Montagem das sub-matrizes Jacobianas B' e B''-----------------*/
	
	matrizes(ce,cle);
	
	/*------------------------------ 
    Elimina a(s) barra(s) slack(s) 
   ------------------------------*/
    elim(ce,1);

    /*-------------------------------------------
    Armazena matriz B'' antes de eliminar barras
    --------------------------------------------*/
    for(i = 0; i < maxnzero; i++) cleb[i] = cle[i];
    
  /*------------------------------
    Elimina as barras slack e PV 
    -----------------------------*/
    elim(cle,0);
	
	/*----------Obtencao de fatores triangulares das matrizes B' e B''--------*/
	redu(ce);
	redu(cle);
	
	/*--------------------Inicio do processo iterativo------------------------*/
	inic = 1;
	
	while(niter <= lim && nitea <= lim && (kp!= 0 || kq != 0)){
		if(inic == 1){
                
	/*---------------CALCULA AS POTENCIAS ATIVAS------------------------------*/
	
	/*-------Calculo das injecoes de potencia nas barras:--- */
		potencia();
		inic = 0;
		
	/*-----	Calculo do MISMATCH ATIVO e do seu valor maximo:----*/
		dpmax = 0.e0;
		for(k = 0; k < nb; k++){
			dx1[k] = 0.e0;
			if(itipo[k] != 2){   // se nao for barra slack
               dx1[k] = pes[k] - p[k];
               dpa = fabs(dx1[k]);
               if(dpa > dpmax){
                  dpmax = dpa;
                  }
               dx1[k] = dx1[k]/v[k];
			   }
		    }
	/*-----Relaxando o criterio de convergencia:----------------------*/
        if(dpmax < 1.e-2 && nitea > 100 ) exit(1);
	        
	/*--------Teste de convergencia (fase ATIVA):---------------*/
		if( dpmax > 1.e+10 ){
            printf("\n Divergencia ...\n");
			exit(1);
            }
		    else
			   if(dpmax <= tola)
				  kp = 0;
		          else{
		/*---------------------------------------------------------------
		Solucao do sistema linear P-TETA para MISMATCH ATIVO:
		Chamada da rotina solu() - Solucao do Sist. Linear B'.dtet = dx1
		----------------------------------------------------------------*/
				      solu(ce,dx1);
				
   	    /*----------------------------------------
		Atualiza o Vetor de Angulos Nodais TETA:
		----------------------------------------*/
				      for(k = 0; k < nb; k++){
                          tet[k] = tet[k] + dx1[k];
                          }
				
		/*------------------------------------------
		Incrementa o Contador de Iteracoes Ativas 
		----------------------------------------- */
                      nitea++;
		/*-------------Retorna a chave REATIVA ao valor inicial-----------*/
		              kq = 1;
                      }  //  fim do else interno
                }     //  fim do else externo
	    else       // else de inic=1
		if(kq == 1){
              
		/*------------------------------------------------------
		OBS: dx1[]-> Armazena os incrementos dos angulos nodais
			dx2[]-> Armazena os incrementos das tensoes nodais
		-------------------------------------------------------- */
	   
       /*-------------- CALCULA AS POTENCIAS REATIVAS ------------------------*/
	
		/*-------Calculo das injecoes de potencia nas barras:--- */
			potencia();
			inic = 1;
		/* ---------------------
        ATUA OU NAO CONTROLES	
		-----------------------*/
		
		/*--1) Teste de Limites de Geracao Reativa----------------------------*/
		    flagq = 0;
			for(k = 0; k < nb; k++){
                 if(itipo[k] == 1 ){
                    qg = q[k] + qca[k];
                    if(qg >= qgmax[k] + 1.e-2) flagq = 999;
                    if(qg <= qgmin[k] - 1.e-2) flagq = 999;
                    }
                  }
            
		/*--2) Teste de Iteracoes para Decidir sobre Controle (etapa reativa)-*/
		    if(niter == nitqm || flagq == 999){
               if(niter >= nitqm){   
                  printf("\n ");
                  printf("\n  Controle PV&Taps - No. Iter Reat niter = %d",niter);
                  printf("\n ");
                  controle(cle,cleb);
                  }
               }
			
		/*---Calculo do MISMATCH REATIVO e do seu valor maximo:-----------*/
			dqmax = 0.e0;
			for(k = 0; k < nb; k++){
				 dx2[k] = 0.e0;
			     if(itipo[k] == 0){   // se nao for barra slack
                    dx2[k] = qes[k] - q[k];
                    dqa = fabs(dx2[k]);
                    if(dqa > dqmax){
                       dqmax = dqa;
                       }
                    dx2[k] = dx2[k]/v[k];
		            }
			     }     // fim do for...
			
            printf("\n Iteracoes Reativas --> %14.10lf  %d",dqmax,niter);
			
		/*-----Relaxando o criterio de convergencia:----------------------*/
	        if(dqmax < 1.e-2 && niter > 100 ) exit(1);

			/*---------Teste de convergencia (fase REATIVA):----------*/
			if(dqmax > 1.e+10 ){
				printf("\n Divergencia ...\n");
				exit(1);
			    }
			else{
				if(dqmax <= tolr)
				   kq = 0;
				   else{
		/*----------Solucao do sistema linear Q-V para MISMATCH REATIVO:
		----------------------------------------------------------------
		Chamada da rotina solu() - Solucao do Sist. Linear B".dv = dx2 
		----------------------------------------------------------------*/
					   solu(cle,dx2);
					
		/*-----------Atualiza o Vetor de Magnitudes Nodais V:-------------*/
					   for(k = 0; k < nb; k++){
					   	   if(itipo[k] == 0){
                              v[k] = v[k] + dx2[k];
                              }
                        }
							
	/*----------Incrementa o Contador de Iteracoes Reativas  ---------*/
					kq = 1;
					niter++;
					
	/*----------Retorna a chave ATIVA ao valor inicial----------*/
					kp = 1;
				}   //  else   (dqmax < tolr)
			}      //  else   (dqmax > 1e.10)
		}         //  else   (se kq=1)
	}            //  while  
	
	ntot = nitea + niter;
 	printf("\n Numero de Iteracoes Ativas + Reativas %4d = \n",ntot);

    /*------Fim do Processo Iterativo por Exceder o Limite Esp. de Iteracoes--*/
    if ((ntot > lim)){
		nconv = 1;
		printf("\n O processo nao converge com %4d iteracoes ...\n", ntot);
		exit(1);
        }
	
	/*---------------- Fim do Processo Iterativo com Sucesso-------------*/
	
/* -------------------------------
   Para acompanhamento da solucao
   -------------------------*/

	  vmi = 1.e+10;
	  vma = 0.e0;
      for(k = 0;k < nb;k++){
          ten = v[k];
          if(ten > vma) kma = k;
          if(ten > vma) vma = ten;
	      if(ten < vmi) kmi = k;
	      if(ten < vmi) vmi = ten;
          }
	  printf("\n %4d %9.4lf %4d %9.4lf \n",nex[kma],v[kma],nex[kmi],v[kmi]);
 
	return;
}
/*--------------------------- Fim de estado() --------------------------------*/

File: estagio/conting/PIPELINE/estado.f
c23456781234567812345678123456781234567812345678123456781234567812345678
c   1       2       3       4       5       6       7       8       9
c ----------------------------------------------------------------------
c
c ----------------------------------------------------------------------
c Subrotina para processar as iteracoes do calculo de fluxo de carga AC
c -----------FAZ CHAMADAS RELATIVAS AS ROTINAS DE ZOLLENKOPF------------

c
c FONTE DESTE ALGORITMO: Fluxograma da pag. 95 do Livro do A. Monticelli
c


      subroutine estado(ce,cle)           ! chamada no programa principal
c
      include 'globvar.inc'
c
      dimension  ce(max),cle(max)
      dimension  dx1(nbmax),dx2(nbmax)
c
      intrinsic DMAX1
     

c ----------------------------------
c Inicializando contadores e chaves
c ----------------------------------

      nitea = 0            ! contadores de iteracoes
      niter = 0
      ntot  = 0

      kp = 1           ! chaves
      kq = 1


     
c ---------------------------------------------------------
c Montando as Sub-matrizes Jacobianas B' e B'' (Versao BX)
c ---------------------------------------------------------
   
c
c Formando a matriz B'
c   

      call matb1(ce)
      nn = 1
      call elim(nn,ce)

c
c Formando a matriz B''
c 

      call matb2(cle)
      nn = 0
      call elim(nn,cle)

      

c ------------------------------------------------------------------
c Obtendo Fatores Triangulares das Sub-matrizes Jacobianas B' e B''
c ------------------------------------------------------------------

c
c  Chamada da rotina redu() - Fatoracao LDU
c
      call redu(ce)




c
c  Chamada da rotina redu() - Fatoracao LDU
c
      call redu(cle)





c ----------------------------
c Inicia o processo iterativo
c ----------------------------

 333  continue


c -----------------
c ITERACOES ATIVAS
c -----------------

c == Calculo das injecoes de potencia nas barras

 203  continue
      call potencia 



c == Calculo do MISMATCH ATIVO e do seu valor maximo 

cc      write(*,*) 
cc      write(*,*)' Iteracao Ativa No. ',nitea

      dpmax = 0.0
      do k = 1,nb
         dx1(k) = pes(k) - p(k) 
         
cc         write(*,*)' dx1(',k,') = ',dx1(k)
 
         dp = ABS(dx1(k))
         dx1(k) = dx1(k)/v(k)
         if( itipo(k) .ne. 2 ) dpmax = DMAX1(dpmax,dp)
      enddo 



c == Teste de convergencia (fase ATIVA)

      if( dpmax .gt. 1.d+10 ) then
          write(*,*)' DIVERGENCIA !!'
      endif

      if( dpmax .le. tol ) then
          kp = 0
          if( kq .eq. 0 ) then
              goto 1000           ! encontrada solucao 
          else
              goto 1050
          endif         
      endif


c == Teste de iteracoes (fase ATIVA) 

      if( nitea .gt. lim ) goto 999 

 
c == Solucao do sistema linear P-TETA para MISMATCH ATIVO    
     
c
c  Chamada da rotina solu() - Solucao do Sist. Linear B'.dtet = dx1
c

      call solu(ce,dx1)


c == Atualiza o Vetor de Angulos Nodais TETA 
     
      
      do k = 1,nb
         tet(k) = tet(k) + dx1(k)
         t(k) = tet(k)
      enddo 

  
c == Incrementa o Contador de Iteracoes Ativas

      nitea = nitea + 1

cc      write(*,*)' Atualiza Iteracao Ativa No. ',nitea

      
c == Retorna a chave REATIVA ao valor inicial

      kq = 1
      

     
   
 1050 continue
   
c -------------------
c ITERACOES REATIVAS
c -------------------  


c == Calculo das injecoes de potencia nas barras

 303  continue
      call potencia 



c == Calculo do MISMATCH REATIVO e do seu valor maximo 

cc      write(*,*) 
cc      write(*,*)' Iteracao Reativa No. ',niter

      dqmax = 0.0
      do k = 1,nb
         dx2(k) = qes(k) - q(k)

cc         write(*,*)' dx2(',k,') = ',dx2(k)        

         dq = ABS(dx2(k))
         if( itipo(k) .le. 0 ) then
             dqmax = DMAX1(dqmax,dq)
             dx2(k) = dx2(k)/v(k) 
         endif
      enddo 



c == Teste de convergencia (fase REATIVA)

      if( dqmax .gt. 1.d+10 ) then
          write(*,*)' DIVERGENCIA !!'
      endif

      if( dqmax .le. tol ) then
          kq = 0
          if( kp .eq. 0 ) then
              goto 1000              ! encontrada solucao 
          else
              goto 2050
          endif         
      endif


c == Teste de iteracoes (fase REATIVA) 

      if( niter .gt. lim ) goto 999 

 
c == Solucao do sistema linear Q-V para MISMATCH REATIVO    
     
c
c  Chamada da rotina solu() - Solucao do Sist. Linear B''.dv = dx2
c

      call solu(cle,dx2)




c == Atualiza o Vetor de Magnitudes Nodais V 
     
      
      do k = 1,nb
         v(k) = v(k) + dx2(k)
      enddo 

  
c == Incrementa o Contador de Iteracoes Reativas

      kq = 1             ! Baleeiro estudou isto em 01/09/95
      niter = niter + 1

cc      write(*,*)' Atualiza Iteracao Reativa No. ',niter



c == Retorna a chave ATIVA ao valor inicial

      kp = 1
 
  
     
c --------------------------------------
c Volta ao Inicio do Processo Iterativo
c --------------------------------------

 2050 continue
      goto 333 
 

c -----------------------------------------------------------------
c Fim do Processo Iterativo por Exceder o Limite Esp. de Iteracoes
c -----------------------------------------------------------------

  999 continue
      ntot = nitea + niter
      nconv = 1
      write(*,555)ntot
  555 format(5x,' O Processo Nao CONVERGE com ',i4,' Meias Iteracoes!!')

      
      
c --------------------------------------
c Fim do Processo Iterativo com Sucesso
c -------------------------------------- 
     
 1000 continue   
      return 
      end               ! Fim da subrotina estado

